src/option.rs

```diff
@@ -76,14 +76,46 @@ where
 
         match keyword.to_ascii_lowercase().as_str() {
             "user" => args.map_owned(User),
-            "port" => args.map_next(|arg, _| Ok(Port(arg.parse()?))),
+            "port" => args.map_next(|arg, _| Ok(Port(a2port(arg)?))),
             "hostname" => args.map_owned(Hostname),
 
             _ => Err(Error::from(DetailedError::BadOption(keyword.to_string()))),
         }
     }
 }
 
+#[cfg(not(feature = "with_libc"))]
+fn a2port(s: &str) -> Result<u16, std::num::ParseIntError> {
+    s.parse()
+}
+
+// See: https://github.com/openssh/openssh-portable/blob/e073106f370cdd2679e41f6f55a37b491f0e82fe/misc.c#L414-L432
+#[cfg(feature = "with_libc")]
+fn a2port(s: &str) -> Result<u16, std::num::ParseIntError> {
+    use libc::getservbyname;
+    use std::convert::TryInto;
+    use std::ffi::CString;
+
+    match s.parse() {
+        res @ Ok(_) => res,
+        num_err @ Err(_) => match CString::new(s) {
+            Ok(cstr) => unsafe {
```

Comment:
Can we shrink the unsafe block?

Reply:
Yup: 20814ef

Switched to two smaller blocks that each depend on different safety properties: the first that the arguments to `getservbyname` are not out of spec, and the second that the pointer it returns is not null.

I *think* these are entirely local properties, and so deserve their own unique safety "arenas."